File: tools/misc/move-python-reference/move_xml.py
"""Converts a MOVE graph to and from an editable XML document.

An interchange format, not one the game loads - the same relationship .fcb has with
Gibbed's XML. Export it, edit it, build it back to a binary .bin.

It is not the engine's own authoring format either. movemgrnamed.bin is closer to
that, but it addresses objects by GUID rather than by stream position and no shipped
executable can read it, so it cannot be the basis for something that builds a
loadable file. What this borrows from the engine is the vocabulary: every field name
is the debug string the matching Transfer call passes.

Pass --names with a *named* twin and criteria are labelled with the channel and enum
value they test. Those labels are informational; the builder ignores them.

  move_xml.py export dlc1.bin dlc1.xml --names movemgrnamed.bin
  move_xml.py build  dlc1.xml out.bin
  move_xml.py verify dlc1.bin --names movemgrnamed.bin
"""
import argparse
import struct
from xml.etree import ElementTree
from xml.sax.saxutils import quoteattr

import move_codec as mc

ROOT = "MoveGraph"
PRINTABLE = set(range(0x20, 0x7F))


def _text(b):
    """A string field as readable text, or None if it is not clean ASCII."""
    return b.decode("ascii") if b and all(c in PRINTABLE for c in b) else None


def to_xml(mf, channels=None):
    out = ['<?xml version="1.0" encoding="utf-8"?>']
    out.append('<%s type=%s version="%d" flags="0x%08X">'
               % (ROOT, quoteattr(struct.pack("<I", mf.mtype).decode("latin1").rstrip("\0")),
                  mf.mversion, mf.flags))
    for kind, name, value in mf.root.ops:
        _emit(out, kind, name, value, 1, channels or [], None)
    out.append("</%s>" % ROOT)
    out.append("")
    return "\n".join(out)


def _annotate(channels, channel, name, value):
    """Informational channel/enum names; the builder ignores them."""
    if not channels:
        return ""
    if name == "m_eValueID" and value < len(channels):
        return " channel=%s" % quoteattr(channels[value][0])
    if name == "m_Value" and channel is not None and channel < len(channels):
        values = channels[channel][1]
        if values and 0 <= value < len(values):
            return " enum=%s" % quoteattr(values[value])
    return ""


def _emit(out, kind, name, value, depth, channels, channel):
    pad = "  " * depth
    n = quoteattr(name)
    if kind == "pnew":
        out.append("%s<obj n=%s class=%s id=\"%d\">"
                   % (pad, n, quoteattr(value.cls), value.index))
        inner = mc.field(value, "m_eValueID")
        for k2, n2, v2 in value.ops:
            _emit(out, k2, n2, v2, depth + 1, channels, inner)
        out.append("%s</obj>" % pad)
    elif kind == "pref":
        out.append('%s<ref n=%s id="%d"/>' % (pad, n, value.index))
    elif kind == "pnull":
        out.append("%s<null n=%s/>" % (pad, n))
    elif kind == "nover":
        out.append("%s<nover n=%s/>" % (pad, n))
    elif kind == "ver":
        out.append('%s<ver n=%s v="%d"/>' % (pad, n, value))
    elif kind == "f32":
        out.append("%s<f32 n=%s %s/>" % (pad, n, _float_attr(value)))
    elif kind in ("u8", "u32", "s32"):
        out.append('%s<%s n=%s v="%d"%s/>'
                   % (pad, kind, n, value, _annotate(channels, channel, name, value)))
    elif kind == "str":
        text = _text(value)
        body = "v=%s" % quoteattr(text) if text is not None else 'hex="%s"' % value.hex()
        out.append("%s<str n=%s %s/>" % (pad, n, body))
    else:
        out.append('%s<%s n=%s hex="%s"/>' % (pad, kind, n, value.hex()))


def _float_attr(raw):
    """Prefer readable decimal, fall back to hex when it would not round-trip."""
    value = struct.unpack("<f", raw)[0]
    text = repr(value)
    try:
        if struct.pack("<f", float(text)) == raw:
            return 'v="%s"' % text
    except (OverflowError, ValueError):
        pass
    return 'hex="%s"' % raw.hex()


def from_xml(text):
    root = ElementTree.fromstring(text)
    if root.tag != ROOT:
        raise mc.Drift("expected a <%s> document, got <%s>" % (ROOT, root.tag))
    mf = mc.MoveFile()
    raw_type = root.get("type", "MVM").encode("latin1")
    mf.mtype = struct.unpack("<I", raw_type.ljust(4, b"\0"))[0]
    mf.mversion = int(root.get("version"))
    mf.flags = int(root.get("flags"), 0)
    mf.root = mc.Obj("#file")
    by_id = {}
    for child in root:
        mf.root.ops.append(_parse(child, by_id))
    mf.seq = [by_id[k] for k in sorted(by_id)]
    return mf


def _parse(el, by_id):
    name = el.get("n", "")
    tag = el.tag
    if tag == "obj":
        obj = mc.Obj(el.get("class"))
        obj.index = int(el.get("id"))
        by_id[obj.index] = obj
        for child in el:
            obj.ops.append(_parse(child, by_id))
        return ("pnew", name, obj)
    if tag == "ref":
        return ("pref", name, by_id[int(el.get("id"))])
    if tag == "null":
        return ("pnull", name, None)
    if tag == "nover":
        return ("nover", name, 0)
    if tag == "ver":
        return ("ver", name, int(el.get("v")))
    if tag == "f32":
        hexed = el.get("hex")
        raw = bytes.fromhex(hexed) if hexed else struct.pack("<f", float(el.get("v")))
        return ("f32", name, raw)
    if tag in ("u8", "u32", "s32"):
        return (tag, name, int(el.get("v")))
    if tag in ("str", "data", "raw"):
        hexed = el.get("hex")
        raw = bytes.fromhex(hexed) if hexed is not None else el.get("v").encode("ascii")
        return (tag, name, raw)
    raise mc.Drift("unexpected element <%s>" % tag)


def main():
    ap = argparse.ArgumentParser(description=__doc__,
                                 formatter_class=argparse.RawDescriptionHelpFormatter)
    ap.add_argument("mode", choices=("export", "build", "verify"))
    ap.add_argument("src")
    ap.add_argument("dst", nargs="?")
    ap.add_argument("--names", metavar="NAMEDBIN",
                    help="a *named* twin to label channels and enum values with")
    args = ap.parse_args()
    channels = mc.channel_table(args.names) if args.names else None

    if args.mode == "build":
        open(args.dst, "wb").write(mc.save(from_xml(open(args.src, encoding="utf-8").read())))
        return
    xml = to_xml(mc.load(args.src), channels)
    if args.mode == "export":
        open(args.dst, "w", encoding="utf-8").write(xml)
        return
    original = open(args.src, "rb").read()
    rebuilt = mc.save(from_xml(xml))
    print("%s" % args.src)
    print("   %d bytes -> %d chars of XML -> %d bytes" % (len(original), len(xml), len(rebuilt)))
    print("   bin -> xml -> bin: %s"
          % ("byte-identical" if rebuilt == original else "*** MISMATCH ***"))


if __name__ == "__main__":
    main()

File: tools/JackAll/src/JackAll.Core/Format/Move/MoveXml.cs
using System.Buffers.Binary;
using System.Globalization;
using System.Text;
using System.Xml;

namespace JackAll.Core.Format.Move;

/// <summary>Converts a MOVE graph to and from an editable XML document.</summary>
/// <remarks>
/// An interchange format, not one the game loads - the same relationship <c>.fcb</c> has with the
/// XML Gibbed's converter produces.
///
/// It is deliberately not <c>movemgrnamed.bin</c>, the engine's own authoring form. That addresses
/// objects by GUID rather than by stream position and no shipped executable can read it, so it
/// cannot be the basis for something that builds a loadable file. What this borrows from the engine
/// is the vocabulary: every field name is the debug string the matching Transfer call passes.
///
/// Both directions stream, because the base graph is 25 MB of XML.
/// See docs/docs/file-formats/move.md.
/// </remarks>
public static class MoveXml
{
    private const string RootName = "MoveGraph";

    /// <summary>Renders a binary MOVE graph as XML.</summary>
    public static string Decode(byte[] data, MoveLabels? labels = null)
        => ToXml(MoveCodec.Load(data), labels);

    /// <summary>Builds a document produced by <see cref="Decode"/> back into a binary graph.</summary>
    public static byte[] Encode(string xml) => MoveCodec.Save(FromXml(xml));

    public static string ToXml(MoveFile file, MoveLabels? labels = null)
    {
        StringBuilder text = new();
        XmlWriterSettings settings = new() { Indent = true, IndentChars = "  " };
        using (XmlWriter writer = XmlWriter.Create(text, settings))
        {
            writer.WriteStartElement(RootName);
            writer.WriteAttributeString("type", FourCc(file.Type));
            writer.WriteAttributeString("version", file.Version.ToString(CultureInfo.InvariantCulture));
            writer.WriteAttributeString("flags", "0x" + file.Flags.ToString("X8"));
            foreach (MoveOp op in file.Root.Ops)
            {
                Write(writer, op, labels, null);
            }

            writer.WriteEndElement();
        }

        return text.ToString();
    }

    private static void Write(
        XmlWriter writer, MoveOp op, MoveLabels? labels, uint? channel)
    {
        switch (op.Kind)
        {
            case MoveOpKind.PointerNew:
                MoveObject target = op.Target!;
                writer.WriteStartElement("obj");
                writer.WriteAttributeString("n", op.Name);
                writer.WriteAttributeString("class", target.ClassName);
                writer.WriteAttributeString("id", target.Index.ToString(CultureInfo.InvariantCulture));
                uint? inner = target.Field("m_eValueID");
                foreach (MoveOp child in target.Ops)
                {
                    Write(writer, child, labels, inner);
                }

                writer.WriteEndElement();
                return;

            case MoveOpKind.PointerRef:
                writer.WriteStartElement("ref");
                writer.WriteAttributeString("n", op.Name);
                writer.WriteAttributeString(
                    "id", op.Target!.Index.ToString(CultureInfo.InvariantCulture));
                writer.WriteEndElement();
                return;

            case MoveOpKind.PointerNull:
                WriteNamed(writer, "null", op.Name);
                return;

            case MoveOpKind.NoVersion:
                WriteNamed(writer, "nover", op.Name);
                return;

            case MoveOpKind.Version:
                WriteValue(writer, "ver", op.Name, op.Number.ToString(CultureInfo.InvariantCulture));
                return;

            case MoveOpKind.F32:
                writer.WriteStartElement("f32");
                writer.WriteAttributeString("n", op.Name);
                WriteFloat(writer, op.Bytes!);
                writer.WriteEndElement();
                return;

            case MoveOpKind.U8:
            case MoveOpKind.U32:
                WriteInteger(writer, op, op.Number.ToString(CultureInfo.InvariantCulture), labels, channel);
                return;

            case MoveOpKind.S32:
                WriteInteger(
                    writer, op,
                    unchecked((int)op.Number).ToString(CultureInfo.InvariantCulture), labels, channel);
                return;

            case MoveOpKind.Str:
                writer.WriteStartElement("str");
                writer.WriteAttributeString("n", op.Name);
                if (MoveText.Printable(op.Bytes!) is { } clean)
                {
                    writer.WriteAttributeString("v", clean);
                }
                else
                {
                    writer.WriteAttributeString("hex", Convert.ToHexString(op.Bytes!));
                }

                writer.WriteEndElement();
                return;

            default:
                writer.WriteStartElement(op.Kind == MoveOpKind.Data ? "data" : "raw");
                writer.WriteAttributeString("n", op.Name);
                writer.WriteAttributeString("hex", Convert.ToHexString(op.Bytes!));
                writer.WriteEndElement();
                return;
        }
    }

    private static void WriteNamed(XmlWriter writer, string element, string name)
    {
        writer.WriteStartElement(element);
        writer.WriteAttributeString("n", name);
        writer.WriteEndElement();
    }

    private static void WriteValue(XmlWriter writer, string element, string name, string value)
    {
        writer.WriteStartElement(element);
        writer.WriteAttributeString("n", name);
        writer.WriteAttributeString("v", value);
        writer.WriteEndElement();
    }

    private static void WriteInteger(
        XmlWriter writer, MoveOp op, string value,
        MoveLabels? labels, uint? channel)
    {
        writer.WriteStartElement(op.Kind.ToString().ToLowerInvariant());
        writer.WriteAttributeString("n", op.Name);
        writer.WriteAttributeString("v", value);
        Annotate(writer, op, labels, channel);
        writer.WriteEndElement();
    }

    /// <summary>Informational channel and enum labels; <see cref="FromXml"/> ignores them.</summary>
    private static void Annotate(
        XmlWriter writer, MoveOp op, MoveLabels? labels, uint? channel)
    {
        if (labels is null)
        {
            return;
        }

        if (labels.PathOf(op.Name, op.Number) is { } path)
        {
            writer.WriteAttributeString("path", path);
            return;
        }

        IReadOnlyList<MoveChannel>? channels = labels.Channels;
        if (channels is null)
        {
            return;
        }

        if (op.Name == "m_eValueID" && op.Number < channels.Count)
        {
            writer.WriteAttributeString("channel", channels[(int)op.Number].Name);
            return;
        }

        if (op.Name != "m_Value" || channel is not { } id || id >= channels.Count)
        {
            return;
        }

        IReadOnlyList<string>? values = channels[(int)id].Values;
        int index = unchecked((int)op.Number);
        if (values is not null && index >= 0 && index < values.Count)
        {
            writer.WriteAttributeString("enum", values[index]);
        }
    }

    /// <summary>Readable decimal when it round-trips, hex when it would not.</summary>
    private static void WriteFloat(XmlWriter writer, byte[] raw)
    {
        float value = BitConverter.ToSingle(raw);
        string text = value.ToString("R", CultureInfo.InvariantCulture);
        if (float.TryParse(text, NumberStyles.Float, CultureInfo.InvariantCulture, out float parsed)
            && BitConverter.GetBytes(parsed).AsSpan().SequenceEqual(raw))
        {
            writer.WriteAttributeString("v", text);
            return;
        }

        writer.WriteAttributeString("hex", Convert.ToHexString(raw));
    }

    private static string FourCc(uint value)
    {
        Span<byte> buffer = stackalloc byte[4];
        BinaryPrimitives.WriteUInt32LittleEndian(buffer, value);
        return Encoding.Latin1.GetString(buffer).TrimEnd('\0');
    }

    public static MoveFile FromXml(string xml)
    {
        MoveFile file = new();
        Dictionary<int, MoveObject> byId = [];
        Stack<MoveObject> stack = new();
        MoveObject current = file.Root;

        using XmlReader reader = XmlReader.Create(
            new StringReader(xml), new XmlReaderSettings { IgnoreWhitespace = true });

        while (reader.Read())
        {
            if (reader.NodeType == XmlNodeType.EndElement)
            {
                if (reader.Name == "obj")
                {
                    current = stack.Pop();
                }

                continue;
            }

            if (reader.NodeType != XmlNodeType.Element)
            {
                continue;
            }

            if (reader.Name == RootName)
            {
                ReadHeader(reader, file);
                continue;
            }

            string name = reader.GetAttribute("n") ?? string.Empty;
            if (reader.Name == "obj")
            {
                MoveObject created = new(reader.GetAttribute("class")
                    ?? throw new MoveFormatException("an <obj> has no class"))
                {
                    Index = int.Parse(reader.GetAttribute("id")!, CultureInfo.InvariantCulture),
                };
                byId[created.Index] = created;
                current.Ops.Add(MoveOp.Pointer(MoveOpKind.PointerNew, name, created));
                if (!reader.IsEmptyElement)
                {
                    stack.Push(current);
                    current = created;
                }

                continue;
            }

            current.Ops.Add(ReadOp(reader, name, byId));
        }

        file.Objects.AddRange(byId.OrderBy(p => p.Key).Select(p => p.Value));
        return file;
    }

    private static void ReadHeader(XmlReader reader, MoveFile file)
    {
        byte[] type = Encoding.Latin1.GetBytes((reader.GetAttribute("type") ?? "MVM").PadRight(4, '\0'));
        file.Type = BinaryPrimitives.ReadUInt32LittleEndian(type);
        file.Version = uint.Parse(reader.GetAttribute("version")!, CultureInfo.InvariantCulture);
        string flags = reader.GetAttribute("flags")!;
        file.Flags = flags.StartsWith("0x", StringComparison.OrdinalIgnoreCase)
            ? Convert.ToUInt32(flags[2..], 16)
            : uint.Parse(flags, CultureInfo.InvariantCulture);
    }

    private static MoveOp ReadOp(XmlReader reader, string name, Dictionary<int, MoveObject> byId)
    {
        string? hex = reader.GetAttribute("hex");
        string? value = reader.GetAttribute("v");
        switch (reader.Name)
        {
            case "ref":
                int id = int.Parse(reader.GetAttribute("id")!, CultureInfo.InvariantCulture);
                return MoveOp.Pointer(MoveOpKind.PointerRef, name, byId[id]);

            case "null":
                return MoveOp.Pointer(MoveOpKind.PointerNull, name, null);

            case "nover":
                return MoveOp.Integer(MoveOpKind.NoVersion, name, 0);

            case "ver":
                return MoveOp.Integer(
                    MoveOpKind.Version, name, uint.Parse(value!, CultureInfo.InvariantCulture));

            case "u8":
                return MoveOp.Integer(
                    MoveOpKind.U8, name, uint.Parse(value!, CultureInfo.InvariantCulture));

            case "u32":
                return MoveOp.Integer(
                    MoveOpKind.U32, name, uint.Parse(value!, CultureInfo.InvariantCulture));

            case "s32":
                return MoveOp.Integer(
                    MoveOpKind.S32, name,
                    unchecked((uint)int.Parse(value!, CultureInfo.InvariantCulture)));

            case "f32":
                byte[] raw = hex is not null
                    ? Convert.FromHexString(hex)
                    : BitConverter.GetBytes(
                        float.Parse(value!, NumberStyles.Float, CultureInfo.InvariantCulture));
                return MoveOp.Blob(MoveOpKind.F32, name, raw);

            case "str":
                return MoveOp.Blob(
                    MoveOpKind.Str, name,
                    hex is not null ? Convert.FromHexString(hex) : Encoding.ASCII.GetBytes(value!));

            case "data":
                return MoveOp.Blob(MoveOpKind.Data, name, Convert.FromHexString(hex!));

            case "raw":
                return MoveOp.Blob(MoveOpKind.Raw, name, Convert.FromHexString(hex!));

            default:
                throw new MoveFormatException($"unexpected element <{reader.Name}>");
        }
    }
}
